### ISSUE TITLE
Float should be float instead of int when it's like 123.0

### ISSUE THREAD

**Issue by tony612** [protobufjs/protobuf.js#1260]
protobuf.js version: 6.8.8

Float field should always be float instead of int. But `123.0` will be converted to `123` using [`Number()`](https://github.com/protobufjs/protobuf.js#valid-message)

```proto
syntax = "proto3";
package test;

message Message {
    float foo1 = 1;
    float foo2 = 2;
}
```

```js
var protobuf = require('protobufjs');

protobuf.load("test.proto", function (err, root) {
  if (err)
    throw err;

  var Message = root.lookupType("test.Message");

  var payload = { foo1: 123.0, foo2: 123.5 };
  var errMsg = Message.verify(payload);
  if (errMsg)
    throw Error(errMsg);
  var message = Message.create(payload);
  var buffer = Message.encode(message).finish();
  var message = Message.decode(buffer);
  console.log(message);
  // Message { foo1: 123, foo2: 123.5 }

});
```

**Comment by iFwu** [comment]
Hello, it's obviously that javascript doesn't have a such `int` type numbr. See below:

![image](https://user-images.githubusercontent.com/16523864/60786186-1ca96b00-a188-11e9-9a0e-8d695ee3c58f.png)

**Comment by tony612** [comment]
But if the value is used as a JSON response, it's different.

**Comment by iFwu** [comment]
No, it's the same. Check out the json specification fot number definition.

**Comment by tony612** [comment]
@iFwu Are they same?
```
number
integer fraction exponent

integer
digit
onenine digits
'-' digit
'-' onenine digits
```
The problem is the receivers will have a problem, like Android clients probably can't handle this when 123 is returned instead of 123.0. We can't say they are the same just because Javascript doesn't have a good type system.

**Comment by iFwu** [comment]
你都定义了float安卓怎么可能收到int？JS数字就是IEEE 754标准的，然后你在proto里定义了什么类型就是什么类型。

**Comment by tony612** [comment]
@iFwu Please see my code above, `foo1` is int after decoding.

**Comment by iFwu** [comment]
你哪看的foo1是int了？`typeof foo1 === 'number'` number就是number

**Comment by tony612** [comment]
@iFwu But if you pass foo1 to the response  clients will receive int, right? Like
```
{foo: foo1} # This is returned as JSON response
```

**Comment by tony612** [comment]
Let me describe more about the situation. We have a nodejs HTTP server, which calls a gRPC server. In the Nodejs server, the response received from gRPC will be used to returned to HTTP client. In this case, clients like Android will receive int instead of float.

**Comment by iFwu** [comment]
> Let me describe more about the situation. We have a nodejs HTTP server, which calls a gRPC server. In the Nodejs server, the response received from the gRPC server will be used to returned to HTTP client. In this case, clients like Android will receive int instead of float.

So there are actual two problems in your situation:
1. When the protobuf float field decoded in JavaScript, the meta type information `float` lost.
2. When send response using Node.js and json format, you can't ensure the numeric type to be correct. (JSON doesn't have a difference between `int` and `float`)

Obviously, you cannot expect `protobuf.js` to solve your problems. The `protobuf.js` lib aims to encode data from JS and decode data from protobuf binary. And all of the jobs are done correctly.

IMO, you can still use protobuf to solve the problem 2, just simply replace the response format from `json` 
 to `protobuf`. And you will get the correct type in Android client.

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
